# spyglass
`spyglass` is a PWM (Position Weight Matrix)-based implementaion of a DNA binding motif finder. It is a tool that identifies enriched motifs in 1D genomic regions given a list of known motifs. 

[Prerequisites](#prerequisites) | [Installation](#install) | [Basic Usage](#usage) | [spyglass Options](#options) | [File formats](#formats) | [Testing](#testing)

<a name="prerequisites"></a>
## Prerequisites
`spyglass` requires the following python libraries to be installed:
- numpy
- pandas
- pyfaidx
- scipy
- seqlogo

These packages can be installed with `pip`:
```
pip install numpy pandas pyfaidx scipy seqlogo
```



<a name="install"></a>
## Installation
Once required libraries are installed (please see [Prerequisites](#prerequisites)), please install `spyglass` with the following commands:
```
git clone https://github.com/kyrafetter/spyglass.git
cd spyglass
python setup.py install
```
Note: if you do not have root access, you can run the command above with additional options to install locally:
```
python setup.py install --user
```
If the install was successful, type `spyglass --help` to show a useful message.

<a name="usage"></a>
## Basic Usage 
The basic usage of `spyglass` is:
```
spyglass [other options] ref.fa peaks.bed motifs.pwm
```
To run `spyglass` on our mini test files (see `example_files`):
```
spyglass test_ref.fa test_peaks.bed test_motifs.pwm
```
This should produce the output below:
```
INSERT OUTPUT
```
To compare to output of Homer `findMotifsGenome.pl`, run:
```
findMotifsGenome.pl test_peaks.bed test_ref.fa ~/example_files/peakAnalysis
```
<a name="options"></a>
## spyglass Options
`spyglass` has the following required arguments:
- `ref.fa`: reference genome in Fasta format
- `peaks.bed`: genomic regions file, for example, peak calls from ChIP-seq datasets.
-  `motifs.pwm`: PWMs of all motifs-of-interest. `spyglass` will determine whether these motifs are significantly enriched in `regions.bed`

Additionally, users may choose to specify the optional options below:
 - `-b BACKGROUND`, `--background BACKGROUND`: BED file of user-specified background genomic peak regions. By default, background sequences will be chosen randomly from the reference genome.
 - `-o FILE`, `--output FILE`: Write output to this file. By default, output is written to stdout.
 - `-l LOGFILE`, `--logfile LOGFILE`: Write input file paths used and runtime to this file. 
 - `-p PVAL`, `--pval PVAL`: P-value threshold to be compared with Fisher exact test p-value for significant enrichment. 
 - `-r REVERSE`, `--reverse REVERSE`: Consider reverse complement in enrichment analysis. 
 - `-s SEQLOGO`, `--seqlogo SEQLOGO`: Generate motif logo of enriched motifs. 
 - `--version VERSION`: Print the version and quit. 

<a name="formats"></a>
## File Formats
`peaks.bed` is a tab-delimited file in BED format with no header. It contains six columns as follows:
```
chromosome start_coordinate  end_coordinate peak_ID . strand(+/-)
```
`ref.fa` is a reference genome sequence in FASTA format. It contains the name of the chromosome followed by the sequence:
```
>chr[name]
[chromosome sequence]
```
`spyglass_results.txt`, the final output file, contains the following tab-delimited columns:
```
motif_name	number_foreground_peaks_with_motif   number_background_peaks_with_motif   p-value   enriched(yes/no)
```
The list of motifs will be sorted so that enriched motifs are listed first and by p-value significance (smallest to largest). 

<a name="testing"></a>
## Testing

<a name="contributors"></a>
## Contributors 
This repository was generated by Michael Chan, Kyra Fetter, and Jessica Wang. We credit [`mypileup`](https://github.com/gymreklab/cse185-demo-project) for inspiration and direction.

Please submit a pull request with any corrections or suggestions. Thanks!
